### PR TITLE
feat: placeos cli

### LINKED ISSUE #38: Runtime environment overridden by `.env`
**Describe the bug**
Environment variables declared within `.env` override the state of the runtime environment within the install script (and likely others). This is the inverse of normal `.env` file behaviour, including within docker-compose (runtime env overrides `.env`).

**To Reproduce**
```bash
$ export PLACE_DOMAIN=example.com
$ ./install

...

=== Setup complete. Login to https://localhost:8443/backoffice/ ...
```

**Expected behavior**
Environment variables that exist should take precedence over those declared in `.env`.

**Additional context**
Cause: https://github.com/place-labs/partner-environment/blob/47f6ca7e39d1ef90b3eb423cabce3eba6b598c7e/install#L52

### PR DESCRIPTION
**Pull request by caspiano**
Begin a clean up of the partner env script.

- Adds pretty progress spinners lol.
- Logs are written to a file rather than stdout, found under `./logs`
- Adds healthchecks and proper `depends_on` relations between services
- Pretty autogenerated service graph!

Nope. ~~Fixes #38~~
It does however ensure the documented environment is not clobbered.

[See the updated README for more information](https://github.com/place-labs/partner-environment/blob/a235f5b2bc32e2443be9feafae870e73184c8844/README.md)

### Future development

- Make it work without git, this includes updating the environment (may not be worth it?)
- Warn if docker is not installed, give a platform-specific link to download
- Allow selection of platform versions
- Calculate the changelog between the current version and the proposed upgrade (this could also be done in frontend)
- Embed some kind of setup wizard. This could select the type of infrastructure you want to deploy on and guide you through the process.
- Option to open a browser with a viewer for each db

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,8 +1,9 @@
...
-www
-data
-ssl
-.env.*
 *.swp
+.env.*
 .htpasswd*
+.logs
+data
 docker-compose.yml
```
Comment (kimburgess) [inline .gitignore:6]: This probably shouldn't be ignored.

File: README.md [hunk 2]
```diff
@@ -35,31 +37,41 @@ These will need to be installed prior to running `./install`:
...
-2. run the spec, compiled with debug symbols
-3. run `gdb` in a local terminal window
-4. connect to the spec remote debug server `target remote localhost:4444`
-5. to debug the driver, not just the spec use `set detach-on-fork on`
-6. type `continue` to start program execution
-7. type `info inferiors` to see the list of running processes
-8. type `inferior 2` to switch to the driver process
+### `$ ./placeos update`
+
+**TODO**
```
Comment (kimburgess) [inline README.md:73]: Can this one stay off the docs until done?

File: docker-compose.yml [hunk 3]
```diff
@@ -114,6 +120,8 @@ services:
     hostname: auth
     networks:
       placeos:
+    healthcheck:
+      test:  wget --no-verbose -q --spider http://0.0.0.0:8080/auth/authority?health=true
```
Comment (kimburgess) [inline docker-compose.yml:124]: https://github.com/PlaceOS/auth/pull/31

File: docker-compose.yml [hunk 4]
```diff
@@ -322,6 +338,8 @@ services:
     restart: always
     container_name: elastic
     hostname: elastic
+    healthcheck:
+      test: wget -q --no-verbose --tries=1 --spider http://localhost:9200/_cat/health
```
Comment (kimburgess) [inline docker-compose.yml:342]:
```suggestion
      test: docker-healthcheck
```
See https://github.com/blacktop/docker-elasticsearch-alpine/blob/master/7.6/docker-healthcheck

File: placeos [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    # Import environment
+    set -a
+    . ./.env
+    set +a
+
+    echo "░░░ Starting PlaceOS <${PLACEOS_TAG}>"
+    echo "░░░ For detailed logging, run \`tail -f ${logfile}\`"
+
+    PLACE_DOMAIN=${PLACE_DOMAIN:-"localhost:8443"}
+    PLACE_EMAIL=${PLACE_EMAIL:-"support@place.tech"}
```
Comment (kimburgess) [inline placeos:130]: Could we try a pull the current users email from their git config, then (silently) fall back to this?

File: placeos [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        "Failed to configure InfluxDB."
+
+    if [[ $setup_sentry == "true" ]]; then
+        run_or_abort \
+            "${base_path}/scripts/setup-sentry.sh" \
+            "Setting up sentry..." \
+            "Failed to setup sentry"
+
+        echo -e "\n░░░ Sentry setup complete. Login to http://$BASE_DOMAIN on port 8989 with "
+        echo "support@place.tech:development"
```
Comment (kimburgess) [inline placeos:162]: Should this address be hardcoded?

File: placeos [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    fi
+
+    [ ${hard_reset} == "true" ] && hard_reset
+
+    run_or_abort \
+        "${base_path}/scripts/run-init-container" \
+        "Initialising PlaceOS with default domain ($BASE_DOMAIN)..." \
+        "Failed to create user entity."
+
+    PLACEOS_DOMAIN=placeos.local:8443
```
Comment (kimburgess) [inline placeos:174]: I don't think this loopback domain is needed any more. @w-le may be able to confirm.

**Review by kimburgess — COMMENTED** [review]
```
───────────────────████
───────────────────█████
───────────────────██████
───────────────────███████
──────────────────████████
──────────────────████████
─────────────────█████████
────────────────█████████
───────────────█████████
──────────────█████████
──────────────██████████████████
────────────█████████████████████
───────────███████████████████████
████████─██████████████████████████
████████─███████████████████████████
████████─████████████████████████████
████████─████████████████████████████
████████─████████████████████████████
████████─████████████████████████████
████████─███████████████████████████
████████─██████████████████████████
████████─█████████████████████████
████████─████████████████████████
████████─███████████████████████
```

File: .gitignore (hunk 1 above)
Comment (caspiano) [reply .gitignore:6]: This is ignored as a result of the current mutation of the docker-compose. I think once the docker-compose.override.yml method is added, we should default track it.

I’ve been directly adding my changes for now.

File: placeos (hunk 5 above)
Comment (caspiano) [reply placeos:130]: Nice idea

File: placeos (hunk 6 above)
Comment (caspiano) [reply placeos:162]: @viv-4

File: placeos (hunk 7 above)
Comment (w-le) [reply placeos:174]: That's correct, as staff-api now supports not unverified ssl to placeos.

Also staff-api now reads the domain of incoming request URIs, and uses it in the `Host:` header for any requests to PlaceOS API, thus a static loopback domain does not need to be added to PlaceOS just for staff-api's usage anymore (and there was much rejoicing).

https://github.com/place-labs/partner-environment/blob/master/docker-compose.yml#L260-L261

Comment (w-le) [reply placeos:174]: So line 174-178 can now be removed

File: docker-compose.yml (hunk 3 above)
Comment (caspiano) [reply docker-compose.yml:124]: Maybe we should keep it for a few versions as older images won’t have the explicit healthcheck.

Let’s say remove in 2 versions?

File: placeos (hunk 5 above)
Comment (caspiano) [reply placeos:137]: Could be part of a wizard if a `.env.email` is not present, or an explicit email arg present.

File: docker-compose.yml [hunk 8]
```diff
@@ -322,6 +339,8 @@ services:
     restart: always
     container_name: elastic
     hostname: elastic
+    healthcheck:
+      test: docker-healthcheck
```
Comment (caspiano) [inline docker-compose.yml:343]: Looks like CI broke after changing this

**Comment by caspiano** [comment]
WOW.  CI has been failing because the action runner scrubs the `--password` argument... how irritating.

**Comment by caspiano** [comment]
The current behaviour of writing the password is not ideal, but this is not for prod use and is the current behaviour

**Comment by stakach** [comment]
I think this is looking great, I think it's worth merging now and we can adjust with pull requests later

**Pull request merged**